waspc/data/Generator/templates/react-app/src/router/linkHelpers.ts

```diff
@@ -0,0 +1,46 @@
+import type { Params, Search } from "./types";
+
+export function interpolatePath(
+  path: string,
+  params?: Params,
+  search?: Search,
+  hash?: string
+) {
+  const interpolatedPath = params ? interpolatePathParams(path, params) : path
+  const interpolatedSearch = search
+    ? `?${new URLSearchParams(search).toString()}`
+    : ''
+  const interpolatedHash = hash ? `#${hash}` : ''
+  return interpolatedPath + interpolatedSearch + interpolatedHash
+}
+
+function interpolatePathParams(path: string, params: Params) {
+  const startsWithSlash = path.startsWith("/");
+  const interpolatedPath = path
+    .split("/")
+    .map(getPathPartsMapper(params))
+    .filter(isValidPathPart)
+    .join("/");
+  return startsWithSlash ? `/${interpolatedPath}` : interpolatedPath;
+}
+
+function getPathPartsMapper(params: Params) {
+  return function mapPathPart(part: string): string | number | undefined {
+    if (part.startsWith(":")) {
+      const paramName = extractParamNameFromPathPart(part);
+      return params[paramName];
+    }
+    return part;
+  }
+}
```

Comment:
I find this a little complicated. Perhaps we had a misunderstanding, I meant "just put the definition of `mapPathPart` under `interpolatePathParams`.

If you prefer to do it with a higher-order function, that's fine too. I 'd still recommend against it since the "fixed" value (i.e., `params`) is always the same within the call of `interpolatePathParams`. In other words, its closure does not need to support being changed.